packages/flutter/lib/src/widgets/slotted_render_object_widget.dart

```diff
@@ -231,14 +241,43 @@ class SlottedRenderObjectElement<S> extends RenderObjectElement {
     }(), '${widget.runtimeType}.slots must not change.');
     assert(slottedMultiChildRenderObjectWidgetMixin.slots.toSet().length == slottedMultiChildRenderObjectWidgetMixin.slots.length, 'slots must be unique');
 
+    final Map<Key, S> oldKeyedSlots = _keyToSlot;
+    _keyToSlot = <Key, S>{};
+
+    // New slots for keyed children. It shouldn't be updated in place because
+    // two children could swap slots.
+    final Map<S, Element> slotToKeyedChild = <S, Element>{};
+    final List<MapEntry<S, Widget?>> unkeyedWidgets = <MapEntry<S, Widget?>>[];
+
     for (final S slot in slottedMultiChildRenderObjectWidgetMixin.slots) {
-      _updateChild(slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot), slot);
+      final Widget? widget = slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot);
+      final Key? newWidgetKey = widget?.key;
+      if (newWidgetKey != null) {
+        assert(!_keyToSlot.containsKey(newWidgetKey));
+        _keyToSlot[newWidgetKey] = slot;
+        final S fromSlot = oldKeyedSlots[newWidgetKey] ?? slot;
```

Comment:
Why can we just use `slot` as `fromSlot`? Isn't it possible that `slot` contains another keyed widget that might move to another slot behind `slot`? E.g. what if we have 
slots A, B
keyed widget K1, K2
unkeyed widget W

We start out with the following slot occupancy:

A: K1
B: W

In the next frame we want to go to:

A: K2
B: K1

I believe the current code would match (via `updateChild`) K1 with K2, which means K2 is not actually moved to B. Instead, a new element is created for K1 in slot B.